src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.Values.StringSegment.cs

```diff
@@ -0,0 +1,1109 @@
+
+            jsonUtf8.Flush();
+
+            Assert.Equal(0, jsonUtf8.BytesPending);
+            Assert.Equal(expected.Length, jsonUtf8.BytesCommitted);
+            JsonTestHelper.AssertContents('"' + expected, output);
+        }
+
+        [Fact]
+        public static void WriteStringValueSegment_MixEncoding()
+        {
+            {
+                var output = new ArrayBufferWriter<byte>();
+                using var jsonUtf8 = new Utf8JsonWriter(output);
+
+                // Becuase the first code point is a surrogate pair, it will be
+                // saved until the next write to complete it. It is saved in the
+                // original encoding, UTF-16, so it will be 0b1101_1000 0b1101_1000
+                jsonUtf8.WriteStringValueSegment("\uD8D8".AsSpan(), false);
+
+                // Now we write a UTF-8 continuation byte. With the previous partial
+                // state, the whole sequence is 0b110_11000 0b110_11000 0b10_111111.
+                jsonUtf8.WriteStringValueSegment([0b10_111111], true);
+
+                jsonUtf8.Flush();
+
+                // If this is interpreted as UTF-8, the first byte is invalid because
+                // it is a 2-byte start unit but the second byte is not a continuation.
+                // So a replacement character gets written for the first byte. The second and
+                // third units are valid and get written as is. Instead, if this is
+                // handled correctly, two replacement characters will be written.
+
+                JsonTestHelper.AssertContents(@"""\uFFFD\uFFFD""", output);
+            }
+
+            {
+                var output = new ArrayBufferWriter<byte>();
+                using var jsonUtf8 = new Utf8JsonWriter(output);
+
+                // The second UTF-16 code unit, when interpreted as UTF-8, is a continuation,
+                // so if the first and second code units are decoded together, they will
+                // form a valid 3-byte sequence.
+                jsonUtf8.WriteStringValueSegment([0b1110_1111], false);
+                jsonUtf8.WriteStringValueSegment("\u8080".AsSpan(), true);
+
+                jsonUtf8.Flush();
+
+                JsonTestHelper.AssertContents(@"""\uFFFD\u8080""", output);
+            }
+
+            {
+                var output = new ArrayBufferWriter<byte>();
+                using var jsonUtf8 = new Utf8JsonWriter(output);
+                jsonUtf8.WriteStringValueSegment([0b110_11111], false);
+                jsonUtf8.Flush();
+                JsonTestHelper.AssertContents("\"", output);
+
+                // Writing empty UTF-8 sequence will still keep the partial code point
+                jsonUtf8.WriteStringValueSegment(ReadOnlySpan<byte>.Empty, false);
+                jsonUtf8.Flush();
+                JsonTestHelper.AssertContents("\"", output);
+
+                // Writing empty UTF-16 sequence will dump the partial UTF-8 code point
```

Comment:
Curious why that happens. Naively, I would expect that an empty UTF-16 sequence should be a no-op.

Reply:
Whenever encoding changes, we just dump the partial code point regardless of the content/size of what's being written (the empty string here), so this behavior is just consistent with that. We can special case this to check if the write is empty and non-final then it can be a no-op. One reasoning for the current behavior is if the user is just writing arbitrary strings then it might be confusing when some of them dump the partial code point and some don't until they track it down to the content.

I consider the whole scenario of switching encodings (especially when there's a partial code point) to be "undefined behavior" from a user's perspective. @GrabYourPitchforks mentioned that our other StreamReader/Writer APIs just misbehave when encodings change across operation.

Reply:
> Whenever encoding changes, we just dump the partial code point regardless of the content/size of what's being written (the empty string here), so this behavior is just consistent with that. 

Sure, it just surprises me that this should happen in the event of a no-op operation. These should usually be just that. I'm not super comfortable with this hybrid mode if I'm honest, it seems to be introducing surprising behavior like this one. I would recommend just disabling it and requiring that the user always specifies the same encoding.

Reply:
Is the fuzzer capable of randomly switching back and forth between encodings for a single string?

Reply:
> Is the fuzzer capable of randomly switching back and forth between encodings for a single string?

Yes, the fuzzer just gets an input of bytes that we can interpret any way we want, so we can implement this encoding switching behavior by just passing a span of the bytes into the writer if we want utf8 or marshalling a span of the bytes into chars and passing that into the writer for utf16. Currently the fuzzer test I added splits the data into 3 segments and does test switching encodings but it's easy enough to make it N splits where N is determined by the input.

> Sure, it just surprises me that this should happen in the event of a no-op call.

In some sense, even if the default values are passed in for the argument (empty input span, isFinal = false), it still isn't a nop call because there is no default for the encoding and the user needs to select one by choosing the overload that they use.

> I would recommend just disabling it and requiring that the user always specifies the same encoding.

Yeah, when encoding switches it will more likely be due to lack of validation or some other error by the caller and we should fail fast instead of silently continuing. If there is a desire for this in the future we should probably make it opt-in through `JsonWriterOptions`.